### PR TITLE
FIX: stale root Figure when adding/updating subfigures

### LINKED ISSUE #28384: [Bug]: subfigure artists not drawn interactively
### Bug summary

When artists are added to a subfigure in interactive mode, they do not appear until I force a draw by resizing the window.

### Code for reproduction

```Python
$ ipython --matplotlib=qt
Python 3.12.3 | packaged by conda-forge | (main, Apr 15 2024, 18:38:13) [GCC 12.3.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.24.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import matplotlib.pyplot as plt

In [2]: fig = plt.figure()

In [3]: sfig1, sfig2 = fig.subfigures(ncols=2)

In [4]: sfig2.suptitle("My Title")
Out[4]: Text(0.5, 0.98, 'My Title')

In [5]: ax = sfig1.subplots()

In [6]: ax.plot([1, 3, 2])
Out[6]: [<matplotlib.lines.Line2D at 0x73a220908ce0>]
```

### Actual outcome

Apparently empty figure
image

### Expected outcome

If I resize the window, the artists appear
image

### Additional information

_No response_

### Operating system

Ubuntu

### Matplotlib Version

`main`

### Matplotlib Backend

QtAgg and TkAgg

### Python version

3.12.3

### Jupyter version

_No response_

### Installation

git checkout

### PR DESCRIPTION
**Pull request by rcomer**
## PR summary

Fixes #28384

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [n/a] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [n/a] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [n/a] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

File: lib/matplotlib/tests/test_figure.py [hunk 1]
```diff
@@ -1741,3 +1741,27 @@ def test_subfigure_row_order():
...
+
+    sfig2 = sfig1.subfigures()
+    assert fig.stale
+
+    fig.draw_without_rendering()
+    assert not fig.stale
+    assert not sfig2.stale
+
+    sfig2.stale = True
+    assert fig.stale
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_figure.py:1767]: Also the intermediate subfigure should change as well?
```suggestion
    fig.draw_without_rendering()
    assert not fig.stale
    assert not sfig1.stale
    assert not sfig2.stale

    sfig2.stale = True
    assert sfig1.stale
    assert fig.stale
```

Comment (rcomer) [reply lib/matplotlib/tests/test_figure.py:1767]: subfigure.figure is the root `Figure`, so the callback doesn't affect the intermediate subfigure

https://github.com/matplotlib/matplotlib/blob/b19794e126c1077b83eba90fe74ab1f67362b1ba/lib/matplotlib/figure.py#L2223

https://github.com/matplotlib/matplotlib/blob/b19794e126c1077b83eba90fe74ab1f67362b1ba/lib/matplotlib/figure.py#L65-L67

@tacaswell thinks that's OK https://github.com/matplotlib/matplotlib/issues/28384#issuecomment-2168364060

Comment (QuLogic) [reply lib/matplotlib/tests/test_figure.py:1767]: Ah, this is more of an interim until #28177?

**Review by QuLogic — APPROVED** [review]

File: lib/matplotlib/tests/test_figure.py (hunk 1 above)
Comment (rcomer) [reply lib/matplotlib/tests/test_figure.py:1767]: Once #28177 is in we can update the callback to explicitly work on the parent figure, so then yes the interim subfigure would also be set stale.  Of course, that would have to wait for 3.10, whereas we can potentially get this fix in for 3.9.1.

**Review by tacaswell — APPROVED** [review]

**Pull request merged**